packages/opds-web-client/src/components/DownloadButton.tsx

```diff
@@ -1,132 +1,99 @@
 import * as React from "react";
 import download from "./download";
+import { useActions } from "./context/ActionsContext";
 
 export interface DownloadButtonProps extends React.HTMLProps<{}> {
   url: string;
   mimeType: string;
   isPlainLink?: boolean;
-  fulfill?: (url: string) => Promise<Blob>;
-  indirectFulfill?: (url: string, type: string) => Promise<string>;
   title?: string;
   indirectType?: string;
 }
 
-/** Shows a button to fulfill and download a book or download it directly. */
-export default class DownloadButton extends React.Component<
-  DownloadButtonProps,
-  {}
-> {
-  constructor(props) {
-    super(props);
-    this.fulfill = this.fulfill.bind(this);
-  }
-
-  render() {
-    // Only get the props needed for the anchor or button element.
-    const {
-      ref,
-      url,
-      mimeType,
-      isPlainLink,
-      fulfill,
-      indirectFulfill,
-      indirectType,
-      title,
-      type,
-      ...props
-    } = this.props;
-
-    return (
-      <span>
-        {this.props.isPlainLink ? (
-          <a
-            className="btn btn-default download-button"
-            {...props}
-            href={this.props.url}
-            target="_blank"
-          >
-            {this.downloadLabel()}
-          </a>
-        ) : (
-          <button
-            className={
-              "btn btn-default download-button download-" +
-              this.fileExtension().slice(1) +
-              "-button"
-            }
-            {...props}
-            onClick={this.fulfill}
-          >
-            {this.downloadLabel()}
-          </button>
-        )}
-      </span>
-    );
-  }
-
-  fulfill() {
-    if (this.isIndirect()) {
-      return this.props
-        .indirectFulfill(this.props.url, this.props.indirectType)
-        .then(url => {
-          window.open(url, "_blank");
-        });
+const DownloadButton: React.FC<DownloadButtonProps> = props => {
+  const {
+    ref,
+    url,
+    mimeType,
+    isPlainLink,
+    indirectType,
+    title,
+    type,
+    ...elementProps
+  } = props;
+  const { actions, dispatch } = useActions();
+  const fulfill = () => {
+    let dispatchFn;
+    let action;
```

Comment:
I would personally just return these directly in the if statements:
```js
if (isIndirect()) {
    const action = ...
    return dispatch(action).then(...)
}
const action = ...
return dispatch(action).then(...)
```

I try to steer clear of having unbound variables that get dynamically bound to things on the whole in order to avoid unexpected undefined cases. Keep in mind that this package doesn't have `strictNullChecks` turned on in `tsconfig`, so typescript won't necessarily catch type errors where those variables don't get defined before they're used

Reply:
I initially tried this but the linter (maybe not Typescript) complained about it. This was the way to get around it. I'll double-check but did find this annoying to deal with.

Reply:
Huh, yea let me know what lint setting or typescript setting complained. Should probably be removed as this would be better

Reply:
Okay, so it's fine. What I couldn't do before was
```
return dispatch(actions.indirectFulfillBook(url, indirectType)).then(
```

Reply:
I would personally still get rid of the `else` block and the `let action`, instead binding action directly. But it's not a huge deal either way